In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Load dataset
df = pd.read_csv('/content/customer_churn_dataset-testing-master.csv')
print("Shape:", df.shape)
print("\nColumn Names:")
print(df.columns.tolist())
print("\nFirst 5 rows:")
df.head()

Shape: (64374, 12)

Column Names:
['CustomerID', 'Age', 'Gender', 'Tenure', 'Usage Frequency', 'Support Calls', 'Payment Delay', 'Subscription Type', 'Contract Length', 'Total Spend', 'Last Interaction', 'Churn']

First 5 rows:


,CustomerID,Age,Gender,Tenure,Usage Frequency,Support Calls,Payment Delay,Subscription Type,Contract Length,Total Spend,Last Interaction,Churn
0,1,22,Female,25,14,4,27,Basic,Monthly,598,9,1
1,2,41,Female,28,28,7,13,Standard,Monthly,584,20,0
2,3,47,Male,27,10,2,29,Premium,Annual,757,21,0
3,4,35,Male,9,12,5,17,Premium,Quarterly,232,18,0
4,5,53,Female,58,24,9,2,Standard,Annual,533,18,0


In [2]:
print("Missing Values:")
print(df.isnull().sum())

print("\nDuplicate Rows:", df.duplicated().sum())

print("\nData Types:")
print(df.dtypes)

# Basic statistics
print("\nBasic Statistics:")
df.describe()

Missing Values:
CustomerID           0
Age                  0
Gender               0
Tenure               0
Usage Frequency      0
Support Calls        0
Payment Delay        0
Subscription Type    0
Contract Length      0
Total Spend          0
Last Interaction     0
Churn                0
dtype: int64

Duplicate Rows: 0

Data Types:
CustomerID            int64
Age                   int64
Gender               object
Tenure                int64
Usage Frequency       int64
Support Calls         int64
Payment Delay         int64
Subscription Type    object
Contract Length      object
Total Spend           int64
Last Interaction      int64
Churn                 int64
dtype: object

Basic Statistics:


,CustomerID,Age,Tenure,Usage Frequency,Support Calls,Payment Delay,Total Spend,Last Interaction,Churn
count,64374.000000,64374.000000,64374.000000,64374.000000,64374.000000,64374.000000,64374.000000,64374.000000,64374.000000
mean,32187.500000,41.970982,31.994827,15.080234,5.400690,17.133952,541.023379,15.498850,0.473685
std,18583.317451,13.924911,17.098234,8.816470,3.114005,8.852211,260.874809,8.638436,0.499311
min,1.000000,18.000000,1.000000,1.000000,0.000000,0.000000,100.000000,1.000000,0.000000
25%,16094.250000,30.000000,18.000000,7.000000,3.000000,10.000000,313.000000,8.000000,0.000000
50%,32187.500000,42.000000,33.000000,15.000000,6.000000,19.000000,534.000000,15.000000,0.000000
75%,48280.750000,54.000000,47.000000,23.000000,8.000000,25.000000,768.000000,23.000000,1.000000
max,64374.000000,65.000000,60.000000,30.000000,10.000000,30.000000,1000.000000,30.000000,1.000000


In [3]:
# 1. Overall Churn Rate
churn_rate = df['Churn'].value_counts()
churn_pct = df['Churn'].value_counts(normalize=True) * 100
print("Churn Count:")
print(churn_rate)
print("\nChurn Percentage:")
print(churn_pct)

# 2. Churn by Subscription Type
sub_churn = df.groupby('Subscription Type')['Churn'].mean() * 100
print("\nChurn Rate by Subscription Type:")
print(sub_churn)

# 3. Churn by Contract Length
contract_churn = df.groupby('Contract Length')['Churn'].mean() * 100
print("\nChurn Rate by Contract Length:")
print(contract_churn)

# 4. Churn by Gender
gender_churn = df.groupby('Gender')['Churn'].mean() * 100
print("\nChurn Rate by Gender:")
print(gender_churn)

# 5. Average Tenure of Churned vs Retained
tenure_churn = df.groupby('Churn')['Tenure'].mean()
print("\nAverage Tenure (0=Retained, 1=Churned):")
print(tenure_churn)

Churn Count:
Churn
0    33881
1    30493
Name: count, dtype: int64

Churn Percentage:
Churn
0    52.631497
1    47.368503
Name: proportion, dtype: float64

Churn Rate by Subscription Type:
Subscription Type
Basic       48.277470
Premium     46.496429
Standard    47.330481
Name: Churn, dtype: float64

Churn Rate by Contract Length:
Contract Length
Annual       46.216721
Monthly      51.608676
Quarterly    44.048190
Name: Churn, dtype: float64

Churn Rate by Gender:
Gender
Female    55.049050
Male      38.579661
Name: Churn, dtype: float64

Average Tenure (0=Retained, 1=Churned):
Churn
0    28.826481
1    35.515200
Name: Tenure, dtype: float64


In [4]:
# 1. Churn by Support Calls
support_churn = df.groupby('Support Calls')['Churn'].mean() * 100
print("Churn Rate by Support Calls:")
print(support_churn)

# 2. Churn by Payment Delay
df['Payment_Delay_Group'] = pd.cut(df['Payment Delay'],
                                    bins=[0,10,20,30],
                                    labels=['Low(0-10)','Medium(10-20)','High(20-30)'])
delay_churn = df.groupby('Payment_Delay_Group')['Churn'].mean() * 100
print("\nChurn Rate by Payment Delay:")
print(delay_churn)

# 3. Churn by Age Group
df['Age_Group'] = pd.cut(df['Age'],
                          bins=[18,30,45,65],
                          labels=['Young(18-30)','Middle(30-45)','Senior(45-65)'])
age_churn = df.groupby('Age_Group')['Churn'].mean() * 100
print("\nChurn Rate by Age Group:")
print(age_churn)

# 4. Revenue Lost due to Churn
churned_revenue = df[df['Churn']==1]['Total Spend'].sum()
retained_revenue = df[df['Churn']==0]['Total Spend'].sum()
print(f"\nRevenue Lost (Churned): ${churned_revenue:,.0f}")
print(f"Revenue Retained: ${retained_revenue:,.0f}")

Churn Rate by Support Calls:
Support Calls
0     24.219851
1     22.884967
2     23.150457
3     24.835909
4     31.837607
5     60.462671
6     60.611538
7     61.782088
8     60.956535
9     60.909364
10    60.695309
Name: Churn, dtype: float64

Churn Rate by Payment Delay:
Payment_Delay_Group
Low(0-10)        10.116756
Medium(10-20)    38.753651
High(20-30)      76.595975
Name: Churn, dtype: float64

Churn Rate by Age Group:
Age_Group
Young(18-30)     44.405976
Middle(30-45)    44.676197
Senior(45-65)    51.041442
Name: Churn, dtype: float64

Revenue Lost (Churned): $15,836,117
Revenue Retained: $18,991,722


/tmp/ipykernel_2911/3201198229.py:10: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  delay_churn = df.groupby('Payment_Delay_Group')['Churn'].mean() * 100
/tmp/ipykernel_2911/3201198229.py:18: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  age_churn = df.groupby('Age_Group')['Churn'].mean() * 100


In [5]:
# Cohort Analysis - Tenure Groups
df['Tenure_Group'] = pd.cut(df['Tenure'],
                    bins=[0,12,24,36,48,60],
                    labels=['0-12 months','13-24 months',
                            '25-36 months','37-48 months',
                            '49-60 months'])

cohort_churn = df.groupby('Tenure_Group')['Churn'].mean() * 100
print("Churn Rate by Tenure Group:")
print(cohort_churn)

# Retention Rate
cohort_retention = 100 - cohort_churn
print("\nRetention Rate by Tenure Group:")
print(cohort_retention)

# Customer Lifetime Value
clv = df.groupby('Churn')['Total Spend'].agg(['mean','sum','count'])
print("\nCustomer Lifetime Value:")
print(clv)

Churn Rate by Tenure Group:
Tenure_Group
0-12 months     31.830986
13-24 months    31.416674
25-36 months    55.681488
37-48 months    56.393184
49-60 months    55.662737
Name: Churn, dtype: float64

Retention Rate by Tenure Group:
Tenure_Group
0-12 months     68.169014
13-24 months    68.583326
25-36 months    44.318512
37-48 months    43.606816
49-60 months    44.337263
Name: Churn, dtype: float64

Customer Lifetime Value:
             mean       sum  count
Churn                             
0      560.541956  18991722  33881
1      519.336143  15836117  30493


/tmp/ipykernel_2911/116669166.py:8: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  cohort_churn = df.groupby('Tenure_Group')['Churn'].mean() * 100


# Customer Retention & Churn Analysis
## Dataset: 64,374 Customer Records

---

## 📊 Key Findings

### 1. Overall Churn Rate
- Total Customers: 64,374
- Churned: 30,493 (47.4%)
- Retained: 33,881 (52.6%)
- **Insight:** Almost half the customers are leaving — urgent action needed!

---

### 2. Churn by Contract Length
- Monthly Plan: 51.6% churn (Highest!)
- Annual Plan: 46.2% churn
- Quarterly Plan: 44.0% churn (Best retention!)
- **Insight:** Promote Quarterly & Annual plans to reduce churn!

---

### 3. Churn by Payment Delay
- Low Delay (0-10 days): 10.1% churn
- Medium Delay (10-20 days): 38.7% churn
- High Delay (20-30 days): 76.6% churn
- **Insight:** Payment delay is the #1 churn predictor!

---

### 4. Churn by Support Calls
- 0-4 calls: ~23-31% churn
- 5+ calls: ~60% churn
- **Insight:** Poor support experience drives customers away!

---

### 5. Churn by Age Group
- Young (18-30): 44.4% churn
- Middle (30-45): 44.7% churn
- Senior (45-65): 51.0% churn
- **Insight:** Senior customers need special attention!

---

### 6. Cohort Analysis — Tenure Groups
- 0-24 months: ~31% churn
- 25-60 months: ~56% churn
- **Insight:** 2-year mark is critical — loyalty programs needed!

---

### 7. Revenue Impact
- Revenue Lost (Churned): $15,836,117
- Revenue Retained: $18,991,722
- **Insight:** Reducing churn = direct revenue growth!

---

## 🎯 Top Business Recommendations

1. **Fix Payment Issues** — Auto-pay reminders & easy payment options
2. **Improve Support** — Faster response, better FAQ
3. **2-Year Loyalty Program** — Special offers at 24-month mark
4. **Senior-Friendly UX** — Dedicated support for 45+ age group
5. **Promote Quarterly Plans** — Better retention than monthly plans